### PR TITLE
Add support for Pandoc `tex_math_single_backslash` and `tex_math_double_backslash` extensions

### LINKED ISSUE #61: Could I get rid of the special meaning of backslash in markdown?
I am using markdown in latex to note faster and I also have the notes in two formats (latex -> pdf and markdown). But it is annoying when I have to use `x\_i` to write `x_i` in latex. I also have to remove all of those slashes when copying to latex format. Could I rerender the underscore `_` in markdown where I want it to have no special meaning? 

I see the option when we can rerender heading like this:
`\markdownSetup{`
`   renderers = {`
`     headingOne = {\section{#1}},`
`     headingTwo = {\subsection{#1}},`
`. },`
`}`

But I don't know how to do in my case, Could you help me?

### LINKED ISSUE #235: Add support for Pandoc `tex_math_single_backslash` syntax extension
We should add support for [Pandoc `tex_math_single_backslash` syntax extension][1].
Progresses #61.

Assigned to @lostenderman as part of grant project [MUNI/33/1658/2022][2].
Tentatively scheduled for milestone 2.22.0 to be delivered by March 31, 2023.

 [1]: https://pandoc.org/MANUAL.html#extension-tex_math_single_backslash
 [2]: https://www.muni.cz/en/research/projects/69762

### LINKED ISSUE #236: Add support for Pandoc `tex_math_double_backslash` syntax extension
We should add support for [Pandoc `tex_math_double_backslash` syntax extension][1].
Progresses #61.

Assigned to @lostenderman as part of grant project [MUNI/33/1658/2022][2].
Tentatively scheduled for milestone 3.0.0 to be delivered by April 30, 2023.

 [1]: https://pandoc.org/MANUAL.html#extension-tex_math_double_backslash
 [2]: https://www.muni.cz/en/research/projects/69762

### PR DESCRIPTION
**Pull request by lostenderman**
Closes #235.
Closes #236.
Closes #61.

### CODE REVIEW

File: tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+\let\infty=\relax
+\let\pi=\relax
+\let\\=\relax
+<<<
+This test ensures that the Lua `texMathDoubleBackslash` and `texMathSingleBackslash` options correctly propagate
+through the plain TeX interface.
+
+\(a\)
+
+\[a\]
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test:17]: This seems to add no value over `tex-math-single-backslash.test`. Remove?

File: tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+interblockSeparator
+inlineMath: a
+interblockSeparator
+displayMath: a
+interblockSeparator
+backslash
+inlineMath: a\\
+interblockSeparator
+backslash
+displayMath: a\\
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test:59]: This is incongruent with the output of Pandoc:

```
$ pandoc -f markdown+tex_math_double_backslash <<< '\\\(a\\\)'
<p>\(a\)</p>
$ pandoc -f markdown+tex_math_double_backslash <<< '\\\[a\\\]'
<p>\[a\]</p>
```

File: markdown.dtx [hunk 3]
```diff
@@ -9225,6 +9225,406 @@ defaultOptions.texMathDollars = false
...
+Then, invoke LuaTeX from the terminal:
+``` sh
+luatex document.tex
+```````
+A PDF document named `document.pdf` should be produced and contain the
+following text:
+
+> \\(E=mc^2\\)
+>
+> \\[\hat{f} \left ( \xi  \right )= \int_{-\infty}^{\infty} f\left ( x  \right ) e^{-i2\pi \xi x} dx\\]
```
Comment (Witiko) [inline markdown.dtx:9280]: Either rewrite to a form that will be recognized by the default extensions of Pandoc (here and hereafter), or add the `tex_math_double_backslash` extension to [`Makefile`][1].

 [1]: https://github.com/Witiko/markdown/blob/36290d5c75687462734a3d7fac29b4bdd90659a5/Makefile#L130

File: markdown.dtx [hunk 4]
```diff
@@ -9225,6 +9225,406 @@ defaultOptions.texMathDollars = false
...
+% \fi
+% \begin{markdown}
+%
+% \Optitem[false]{texMathDoubleBackslash}{\opt{true}, \opt{false}}
+%
+:    true
+
+     :  Enable the Pandoc `tex_math_double_backslash` syntax extension.
+
+        ```
```
Comment (Witiko) [inline markdown.dtx:9241]:
```suggestion
     :  Enable the Pandoc `tex_math_double_backslash` syntax extension:

        ``` md
```

File: markdown.dtx [hunk 5]
```diff
@@ -9225,6 +9225,406 @@ defaultOptions.texMathDollars = false
...
+Then, invoke LuaTeX from the terminal:
+``` sh
+luatex document.tex
+```````
+A PDF document named `document.pdf` should be produced and contain the
+following text:
+
+> \(E=mc^2\)
+>
+> \[\hat{f} \left ( \xi  \right )= \int_{-\infty}^{\infty} f\left ( x  \right ) e^{-i2\pi \xi x} dx\]
```
Comment (Witiko) [inline markdown.dtx:9480]: Either rewrite to a form that will be recognized by the default extensions of Pandoc (here and hereafter), or add the `tex_math_single_backslash` extension to [`Makefile`][1].

 [1]: https://github.com/Witiko/markdown/blob/36290d5c75687462734a3d7fac29b4bdd90659a5/Makefile#L130

File: markdown.dtx [hunk 6]
```diff
@@ -16277,11 +16677,11 @@ following text:
...
 #### Tex Math Renderers
-The \mdef{markdownRendererInlineMath} and \mdef{markdownRendererDisplayMath} macros 
-represent inline and display \TeX{} math. 
+The \mdef{markdownRendererInlineMath} and \mdef{markdownRendererDisplayMath} macros
+represent inline and display \TeX{} math.
 Both macros receive a single argument that corresponds to the tex math content.
-These macros will only be produced, when the \Opt{texMathDollars}
-option is enabled.
+These macros will only be produced, when the \Opt{texMathDollars},
+\Opt{texMathSingleBackslash} or \Opt{texMathDoubleBackslash} option is enabled.
```
Comment (Witiko) [inline markdown.dtx:16684]:
```suggestion
\Opt{texMathSingleBackslash}, or \Opt{texMathDoubleBackslash} option is enabled.
```

File: markdown.dtx [hunk 7]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+
+      local display_math_single = between(C( backslash_math_content),
+                                          display_math_opening_single,
+                                          display_math_closing_single)
+
+      local display_math = parsers.fail
+
+      local inline_math = parsers.fail
+
+      if (tex_math_dollars) then
```
Comment (Witiko) [inline markdown.dtx:26367]:
```suggestion
      if tex_math_dollars then
```

File: markdown.dtx [hunk 8]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+        display_math = display_math + display_math_dollars
+        inline_math = inline_math + inline_math_dollars
+      end
+
+      if (tex_math_double_backslash) then
```
Comment (Witiko) [inline markdown.dtx:26372]:
```suggestion
      if tex_math_double_backslash then
```

File: markdown.dtx [hunk 9]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+
+      local TexMath = display_math / writer.display_math
+                    + inline_math / writer.inline_math
 
       self.insert_pattern("Inline after Emph",
-                          TexMathDollars, "TexMathDollars")
+                          TexMath, "TexMath")
 
-      self.add_special_character("$")
+      if (tex_math_dollars) then
```
Comment (Witiko) [inline markdown.dtx:26388]:
```suggestion
      if tex_math_dollars then
```

File: markdown.dtx [hunk 10]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+        self.add_special_character("$")
+      end
+
+      if (tex_math_single_backslash or tex_math_double_backslash) then
```
Comment (Witiko) [inline markdown.dtx:26392]:
```suggestion
      if tex_math_single_backslash or tex_math_double_backslash then
```

File: markdown.dtx [hunk 11]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
       end
 
-      local inlinemathtail  = B( parsers.any * parsers.nonspacechar
-                               + parsers.backslash * parsers.any)
-                            * parsers.dollar 
-                            * -#(parsers.digit)
+      local allowed_before_closing = B( parsers.backslash * parsers.any
+                                      + parsers.any * (parsers.nonspacechar - parsers.backslash))
+
+      -- tex_math_dollars
```
Comment (Witiko) [inline markdown.dtx:26280]: Instead of an opaque comment `-- tex_math_dollars`, how about a sentence or two in natural language?

```suggestion
%    \end{macrocode}
% \begin{markdown}
%
% The following patterns implement the Pandoc `tex_math_dollars` syntax extension.
%
% \end{markdown}
%  \begin{macrocode}
```

File: markdown.dtx [hunk 12]
```diff
@@ -25827,16 +26227,19 @@ end
 %    \end{macrocode}
 % \begin{markdown}
 %
-%#### Tex Math Dollars
+%#### Tex Math
 %
-% The \luamdef{extensions.tex_math_dollars} function implements the Pandoc
-% tex_math_dollars syntax extension.
+% The \luamdef{extensions.tex_math} function implements the Pandoc
+% tex_math_dollars, tex_math_single_backslash and tex_math_double_backslash
```
Comment (Witiko) [inline markdown.dtx:26233]:
```suggestion
% `tex_math_dollars`, `tex_math_single_backslash` and `tex_math_double_backslash`
```

File: markdown.dtx [hunk 13]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
-                                  - parsers.dollar),
-                                 parsers.dollar * #(parsers.nonspacechar),
-                                 inlinemathtail)
+      local display_math_dollars = between(C( dollar_math_content),
+                                           display_math_opening_dollars,
+                                           display_math_closing_dollars)
 
-      local displaymathdelim  = parsers.dollar 
-                              * parsers.dollar
+      -- tex_math_?_backslash
```
Comment (Witiko) [inline markdown.dtx:26307]: Instead of an opaque comment `-- tex_math_?_backslash`, how about a sentence or two in natural language?

```suggestion
%    \end{macrocode}
% \begin{markdown}
%
% The following patterns implement the Pandoc `tex_math_single_backslash` and
% `tex_math_double_backslash` syntax extensions.
%
% \end{markdown}
%  \begin{macrocode}
```

File: markdown.dtx [hunk 14]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+      local backslash_math_content  = parsers.any 
+                                    - parsers.blankline^2
 
-      local displaymath = between(C( parsers.backslash^-1 
-                                   * parsers.any 
-                                   - parsers.blankline^2 
-                                   - parsers.dollar),
-                                  displaymathdelim,
-                                  displaymathdelim)
+      -- tex_math_double_backslash
```
Comment (Witiko) [inline markdown.dtx:26311]: Instead of an opaque comment `-- tex_math_double_backslash`, how about a sentence or two in natural language?

```suggestion
%    \end{macrocode}
% \begin{markdown}
%
% The following patterns implement the Pandoc `tex_math_double_backslash` syntax extension.
%
% \end{markdown}
%  \begin{macrocode}
```

File: markdown.dtx [hunk 15]
```diff
@@ -25871,35 +26274,128 @@ M.extensions.tex_math_dollars = function()
...
+      local display_math_closing_double = allowed_before_closing
+                                        * parsers.backslash
+                                        * parsers.backslash
+                                        * parsers.rbracket
+
+      local display_math_double = between(C( backslash_math_content),
+                                          display_math_opening_double,
+                                          display_math_closing_double)
+
+      -- tex_math_single_backslash
```
Comment (Witiko) [inline markdown.dtx:26339]: Instead of an opaque comment `-- tex_math_single_backslash`, how about a sentence or two in natural language?

```suggestion
%    \end{macrocode}
% \begin{markdown}
%
% The following patterns implement the Pandoc `tex_math_single_backslash` syntax extension.
%
% \end{markdown}
%  \begin{macrocode}
```

**Review by Witiko — CHANGES_REQUESTED** [review]
Please, also revert the part of https://github.com/Witiko/markdown/commit/5b49318a57b426231c096d295e82867024d62345 that enables `texMathDollars` in the example documents: We don't use math in `examples/example.md`, so adding the option does nothing for the user. Furthermore, even if we had an example of math in `examples/example.md`, we would also need to add `texMathDollars=true` to `examples/Makefile`. I missed this in #267.

Alternatively, please add examples of `tex_math_*` to `example.md` and update `examples/{Makefile,context*.tex,latex.tex}` accordingly.

File: tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test (hunk 2 above)
Comment (lostenderman) [reply tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test:59]: Both extensions are enabled in this unit-test.

```
$ pandoc -f markdown+tex_math_single_backslash+tex_math_double_backslash <<< '\\\(a\\\)'
<p>\<span class="math inline">$a\\$</span></p>
```

or more importantly `tex_math_single_backslash` is enabled.

```
pandoc -f markdown+tex_math_single_backslash <<< '\\\(a\\\)'
<p>\<span class="math inline">$a\\$</span></p>
```

Comment (Witiko) [reply tests/testfiles/lunamark-markdown/tex-math-single-and-double-backslashes.test:36]: Interesting. Perhaps we should add `\\\(a\\\)` and `\\\[a\\\]` to `tex-math-single-backslash.test` and `tex-math-double-backslash.test` as well? Seems like an edge case that you can imagine would parse differently in any combination of the options.

File: examples/example.md [hunk 16]
```diff
@@ -194,4 +194,4 @@ This is inline and display TeX math created using double backslashes:
 
 \\(E=mc^2\\)
 
-\\[E=mc^2\\]
\ No newline at end of file
+\\[E=mc^2\\]
```
Comment (Witiko) [inline examples/example.md:197]: [We normalize the input][1], so that it always ends with a unix-style trailing `\n`. If the reader gives different results with/without trailing `\n`, that should be considered a bug.

 [1]: https://github.com/Witiko/markdown/blob/36290d5c75687462734a3d7fac29b4bdd90659a5/markdown.dtx#L24127-L24137

Comment (lostenderman) [reply examples/example.md:197]: Well, the output is the same, just the MarkdownLink check failed because of it.

```
examples/example.md:197:12 MD047/single-trailing-newline Files should end with a single newline character
```

Comment (Witiko) [reply examples/example.md:197]: Silly MarkdownLint.

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+\let\\=\relax
+<<<
+This test ensures that the Lua `texMathSingleBackslash` option correctly propagates
+through the plain TeX interface.
+
+\(E=mc^2\)
+
+\[\hat{f} \left ( \xi  \right )= \int_{-\infty}^{\infty} f\left ( x  \right ) e^{-i2\pi \xi x} dx\]
+
+Inline or display math may not be empty
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:18]:
```suggestion
Inline or display math may not be empty.
```

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+
+\(\)
+
+\[\]
+
+or have any blank lines
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:24]:
```suggestion
Inline or display math may not have any blank lines:
```

Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:24]:
```suggestion
Inline or display math may not have any blank lines:
```

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+\(a
+
+c\)
+
+\[a
+
+c\]
+
+No spaces are allowed at the beginning or end of the inline math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:34]:
```suggestion
No spaces are allowed at the beginning or end of the inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+
+\( \infty \)
+
+unless they are escaped
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:38]: Escaped spaces are allowed at the beginning or the end of inline math:

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+\(\ \)
+
+Note that this is inline math with two backslashes
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:42]:
```suggestion
The following is inline math with two backslashes:
```

File: tests/testfiles/lunamark-markdown/tex-math-double-backslash.test [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+No spaces are allowed at the beginning or end of the inline math
+
+\\( \infty \\)
+
+unless they are escaped
+
+\\(\ \\)
+
+Note that this is not inline math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-double-backslash.test:42]:
```suggestion
The following is not inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-single-backslash.test (hunk 20 above)
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-single-backslash.test:38]:
```suggestion
Escaped spaces are allowed at the beginning or the end of inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-double-backslash.test [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+\let\\=\relax
+<<<
+This test ensures that the Lua `texMathDoubleBackslash` option correctly propagates
+through the plain TeX interface.
+
+\\(E=mc^2\\)
+
+\\[\hat{f} \left ( \xi  \right )= \int_{-\infty}^{\infty} f\left ( x  \right ) e^{-i2\pi \xi x} dx\\]
+
+Inline or display math may not be empty
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-double-backslash.test:18]:
```suggestion
Inline or display math may not be empty:
```

File: tests/testfiles/lunamark-markdown/tex-math-double-backslash.test [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+
+\\(\\)
+
+\\[\\]
+
+or have any blank lines
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-double-backslash.test:24]:
```suggestion
Inline or display math may not contain any blank lines:
```

File: tests/testfiles/lunamark-markdown/tex-math-double-backslash.test [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+
+\\(a
+
+c\\)
+
+\\[a
+
+c\\]
+
+No spaces are allowed at the beginning or end of the inline math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-double-backslash.test:34]:
```suggestion
No spaces are allowed at the beginning or the end of inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-double-backslash.test [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+
+\\( \infty \\)
+
+unless they are escaped
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-double-backslash.test:38]:
```suggestion
Escaped spaces are allowed at the beginning or the end of inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 27]
```diff
@@ -15,25 +15,25 @@ $E=mc^2$
 
 $$\hat{f} \left ( \xi  \right )= \int_{-\infty}^{\infty} f\left ( x  \right ) e^{-i2\pi \xi x} dx$$
 
-to produce inline and display math both beginning and ending dollar signs must be present
+To produce inline and display math both beginning and ending dollar signs must be present
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:18]:
```suggestion
To produce inline and display math, both beginning and ending dollar signs must be present:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 28]
```diff
@@ -15,25 +15,25 @@ $E=mc^2$
...
 
 $2.56
 
 $$2.56
 
-no spaces are allowed at the beginning or end of the inline math
+No spaces are allowed at the beginning or end of the inline math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:24]:
```suggestion
No spaces are allowed at the beginning or the end of inline math:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 29]
```diff
@@ -15,25 +15,25 @@ $E=mc^2$
...
 
 $\ $
 
-this is not inline math, because the closing dollar sign is followed by a digit
+This is not inline math, because the closing dollar sign is followed by a digit
 
 $a$0
 
-both inline and display math may span multiple lines, but cannot contain a blank line
+Both inline and display math may span multiple lines, but cannot contain a blank line
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:36]:
```suggestion
Both inline and display math may span multiple lines, but cannot contain blank lines:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 30]
```diff
@@ -57,22 +57,22 @@ $$
 
 $$$$
 
-note that this is a display math
+Note that this is a display math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:60]:
```suggestion
The following is display math:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 31]
```diff
@@ -57,22 +57,22 @@ $$
...
 
 $$
 $$
 
-no unescaped dollar signs are allowed 
+No unescaped dollar signs are allowed
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:65]:
```suggestion
Non-escaped dollar signs will always start inline and display math:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 32]
```diff
@@ -57,22 +57,22 @@ $$
...
 
-no unescaped dollar signs are allowed 
+No unescaped dollar signs are allowed 
 
 $.$.$
 
 $$.$.$$
 
-escaped dollar signs are allowed
+Escaped dollar signs are allowed
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:71]:
```suggestion
Escaped dollar signs will not start inline or display math:
```

File: tests/testfiles/lunamark-markdown/tex-math-dollars.test [hunk 33]
```diff
@@ -57,22 +57,22 @@ $$
...
 
 $\$$
 
-note that this is an escaped dollar sign in a display math
+Note that this is an escaped dollar sign in a display math
```
Comment (Witiko) [inline tests/testfiles/lunamark-markdown/tex-math-dollars.test:75]:
```suggestion
The following is an escaped dollar sign inside display math:
```

**Review by Witiko — APPROVED** [review]
